bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java

```diff
@@ -1094,24 +1259,25 @@ private Header getHeaderForLogId(long entryLogId) throws IOException {
     }
 
     private BufferedReadChannel getChannelForLogId(long entryLogId) throws IOException {
-        BufferedReadChannel fc = getFromChannels(entryLogId);
-        if (fc != null) {
-            return fc;
-        }
-        File file = findFile(entryLogId);
-        // get channel is used to open an existing entry log file
-        // it would be better to open using read mode
-        FileChannel newFc = new RandomAccessFile(file, "r").getChannel();
-        FileChannel oldFc = logid2FileChannel.putIfAbsent(entryLogId, newFc);
-        if (null != oldFc) {
-            newFc.close();
-            newFc = oldFc;
+        BufferedReadChannel brc = getFromChannels(entryLogId);
+        if (brc != null) {
+            return brc;
+        }
+        FileChannelBackingCache.CachedFileChannel cachedFileChannel = null;
+        try {
+            do {
+                cachedFileChannel = logid2FileChannel.loadFileChannel(entryLogId);
+            } while (!cachedFileChannel.tryRetain());
+        } finally {
+            if (null != cachedFileChannel) {
+                cachedFileChannel.release();
```

Comment:
?
You're releasing the reference straight away after getting it. By the time cachedFileChannel is used on L1278, the reference could easily be dead.

Reply:
`loadFileChannel` and `tryRetain` both add one reference count, so releasing one is necessary.